### PR TITLE
Feat topics controller

### PR DESCRIPTION
**Pull request by fanatic75**
## What does this PR do?

Adds Topics controller

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Screenshots may also be helpful.)

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [ ] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/controllers/api/messaging.php [hunk 1]
```diff
@@ -0,0 +1,1631 @@
...
+
+        if ($description) {
+            $topic->setAttribute('description', $description);
+        }
+
+        try {
+            $topic = $dbForProject->createDocument('topics', $topic);
+            $response
+                ->setStatusCode(Response::STATUS_CODE_CREATED)
+                ->dynamic($topic, Response::MODEL_TOPIC);
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1322]: Let's move this after the try/catch

File: app/controllers/api/messaging.php [hunk 2]
```diff
@@ -0,0 +1,1631 @@
...
+    ->label('sdk.auth', [APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'updateTopic')
+    ->label('sdk.description', '/docs/references/messaging/update-topic.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_TOPIC)
+    ->param('topicId', '', new UID(), 'Topic ID.')
+    ->param('name', '', new Text(128), 'Topic Name.', true)
+    ->param('description', null, new Text(128), 'Topic Description.', true)
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1343]: This should be 2048 for update too

File: app/controllers/api/messaging.php [hunk 3]
```diff
@@ -0,0 +1,1631 @@
...
+
+        $topic = $dbForProject->deleteDocument('topics', $topicId);
+        $response->noContent();
+    });
+
+App::get('/v1/messaging/topics/:topicId/subscribers')
+    ->desc('List topic\'s subscribers.')
+    ->groups(['api', 'messaging'])
+    ->label('scope', 'subscribers.read')
+    ->label('sdk.auth', [APP_AUTH_TYPE_JWT, APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1398]: This should be ADMIN and KEY only

File: app/controllers/api/messaging.php [hunk 4]
```diff
@@ -0,0 +1,1631 @@
...
+                'subscribers' => $subscribers,
+                'total' => \count($subscribers),
+            ]), Response::MODEL_SUBSCRIBER_LIST);
+    });
+
+App::get('/v1/messaging/topics/:topicId/subscriber/:subscriberId')
+    ->desc('Get a topic\'s subscriber.')
+    ->groups(['api', 'messaging'])
+    ->label('scope', 'subscribers.read')
+    ->label('sdk.auth', [APP_AUTH_TYPE_JWT, APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1430]: This should be ADMIN and KEY only

File: app/controllers/api/messaging.php [hunk 5]
```diff
@@ -0,0 +1,1631 @@
...
+
+App::post('/v1/messaging/topics/:topicId/subscribers')
+    ->desc('Adds a Subscriber to a Topic.')
+    ->groups(['api', 'messaging'])
+    ->label('audits.event', 'subscribers.create')
+    ->label('audits.resource', 'subscribers/{response.$id}')
+    ->label('scope', 'subscribers.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_JWT, APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'addSubscriber')
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1466]: Let's name this `createSubscriber`

File: app/controllers/api/messaging.php [hunk 6]
```diff
@@ -0,0 +1,1631 @@
...
+            'targetId' => $targetId,
+            'targetInternalId' => $target->getInternalId(),
+        ]);
+
+        try {
+            $subscriber = $dbForProject->createDocument('subscribers', $subscriber);
+            $dbForProject->deleteCachedDocument('topics', $topicId);
+            $response
+            ->setStatusCode(Response::STATUS_CODE_CREATED)
+            ->dynamic($subscriber, Response::MODEL_SUBSCRIBER);
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1508]: Let's move this after the try/catch and indent the continuations

Comment (fanatic75) [reply app/controllers/api/messaging.php]: done this

File: app/controllers/api/messaging.php (hunk 5 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: done this

File: app/controllers/api/messaging.php (hunk 4 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: done this

File: app/controllers/api/messaging.php (hunk 2 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: done this

File: app/controllers/api/messaging.php [hunk 7]
```diff
@@ -1192,19 +1192,366 @@
...
+        try {
+            $subscriber = $dbForProject->createDocument('subscribers', $subscriber);
+            $dbForProject->deleteCachedDocument('topics', $topicId);
+        } catch (DuplicateException) {
+            throw new Exception(Exception::SUBSCRIBER_ALREADY_EXISTS);
+        }
+
+        $response
+                ->setStatusCode(Response::STATUS_CODE_CREATED)
+                ->dynamic($subscriber, Response::MODEL_SUBSCRIBER);
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1441]: Indentation

File: tests/e2e/Services/Messaging/MessagingCustomClientTest.php [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Tests\E2E\Services\Messaging;
+
+use Tests\E2E\Scopes\ProjectCustom;
+use Tests\E2E\Scopes\Scope;
+use Tests\E2E\Scopes\SideClient;
+
+class MessagingCustomClientTest extends Scope
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply tests/e2e/Services/Messaging/MessagingCustomClientTest.php:9]: We already have custome server test

File: src/Appwrite/Utopia/Database/Validator/Queries/Providers.php [hunk 9]
```diff
@@ -9,7 +9,9 @@ class Providers extends Base
         'provider',
         'type',
         'default',
-        'enabled'
+        'enabled',
+        'credentials',
+        'options'
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply src/Appwrite/Utopia/Database/Validator/Queries/Providers.php]: removed these

File: src/Appwrite/Utopia/Database/Validator/Queries/Messages.php [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+class Messages extends Base
+{
+    public const ALLOWED_ATTRIBUTES = [
+        'to',
+        'providerId',
+        'deliveredAt',
+        'deliveredTo',
+        'deliveryErrors',
+        'status',
+        'description'
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply src/Appwrite/Utopia/Database/Validator/Queries/Messages.php]: added data here

File: tests/e2e/Services/GraphQL/MessagingTest.php [hunk 11]
```diff
@@ -257,4 +258,405 @@ public function testDeleteProvider(array $providers)
...
+            'sendgrid' => [
+                'providerId' => ID::unique(),
+                'name' => 'Sengrid1',
+                'apiKey' => 'my-apikey',
+            ]
+        ];
+        $query = $this->getQuery(self::$CREATE_SENDGRID_PROVIDER);
+        $graphQLPayload = [
+                'query' => $query,
+                'variables' => $providerParam['sendgrid'],
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply tests/e2e/Services/GraphQL/MessagingTest.php]: fixed these, added more graphql tests for user target too

File: app/controllers/api/messaging.php [hunk 12]
```diff
@@ -1192,19 +1192,366 @@
...
+
+        try {
+            $topic = $dbForProject->createDocument('topics', $topic);
+        } catch (DuplicateException) {
+            throw new Exception(Exception::TOPIC_ALREADY_EXISTS);
+        }
+
+        $response
+                ->setStatusCode(Response::STATUS_CODE_CREATED)
+                ->dynamic($topic, Response::MODEL_TOPIC);
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply app/controllers/api/messaging.php]: done these

File: app/controllers/api/messaging.php [hunk 13]
```diff
@@ -1192,19 +1192,366 @@
...
+    ->action(function (array $queries, Database $dbForProject, Response $response) {
+        $queries = Query::parseQueries($queries);
+
+        // Get cursor document if there was a cursor query
+        $cursor = Query::getByType($queries, [Query::TYPE_CURSORAFTER, Query::TYPE_CURSORBEFORE]);
+        $cursor = reset($cursor);
+
+        if ($cursor) {
+            $topicId = $cursor->getValue();
+            $cursorDocument = Authorization::skip(fn () => $dbForProject->find('topics', [
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply app/controllers/api/messaging.php]: used getDocument

File: app/controllers/api/messaging.php [hunk 14]
```diff
@@ -1192,19 +1192,366 @@
...
+    ->label('sdk.auth', [APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'updateTopic')
+    ->label('sdk.description', '/docs/references/messaging/update-topic.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_TOPIC)
+    ->param('topicId', '', new UID(), 'Topic ID.')
+    ->param('name', '', new Text(128), 'Topic Name.', true)
+    ->param('description', null, new Text(2048), 'Topic Description.', true)
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply app/controllers/api/messaging.php]: set the default here as empty string

File: app/controllers/api/messaging.php [hunk 15]
```diff
@@ -1192,19 +1192,366 @@
...
+    ->inject('dbForProject')
+    ->inject('response')
+    ->action(function (string $topicId, Database $dbForProject, Response $response) {
+        $topic = $dbForProject->getDocument('topics', $topicId);
+
+        if ($topic->isEmpty()) {
+            throw new Exception(Exception::TOPIC_NOT_FOUND);
+        }
+
+        $topic = $dbForProject->deleteDocument('topics', $topicId);
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply app/controllers/api/messaging.php]: deleting all the subscribers using delete worker

File: app/controllers/api/messaging.php (hunk 7 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: fixed

File: app/controllers/api/messaging.php [hunk 16]
```diff
@@ -1216,11 +1563,12 @@
     ->param('content', '', new Text(64230), 'Email Content.')
     ->param('status', 'processing', new WhiteList(['draft', 'processing']), 'Message Status.', true)
     ->param('html', false, new Boolean(), 'Is content of type HTML', true)
+    ->param('deliveryTime', null, new DatetimeValidator(), 'Delivery time for message.', true)
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply app/controllers/api/messaging.php]: updated dateTimeValidator to have a flag called acceptOldDates, default is true

File: app/controllers/api/messaging.php [hunk 17]
```diff
@@ -1192,19 +1191,368 @@
...
+    ->inject('dbForProject')
+    ->inject('response')
+    ->action(function (string $topicId, Database $dbForProject, Response $response) {
+        $topic = Authorization::skip(fn () => $dbForProject->getDocument('topics', $topicId));
+
+        if ($topic->isEmpty()) {
+            throw new Exception(Exception::TOPIC_NOT_FOUND);
+        }
+
+        $subscribers = $dbForProject->find('subscribers', [
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1466]: Can we allow querying here? At least for pagination if not filtering

File: app/controllers/api/messaging.php [hunk 18]
```diff
@@ -1216,11 +1564,12 @@
     ->param('content', '', new Text(64230), 'Email Content.')
     ->param('status', 'processing', new WhiteList(['draft', 'processing']), 'Message Status.', true)
     ->param('html', false, new Boolean(), 'Is content of type HTML', true)
+    ->param('deliveryTime', null, new DatetimeValidator(false), 'Delivery time for message.', true)
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1567]: Let's add the parameter name so it's clear what `false` means here

File: app/controllers/api/messaging.php [hunk 19]
```diff
@@ -1216,11 +1564,12 @@
     ->param('content', '', new Text(64230), 'Email Content.')
     ->param('status', 'processing', new WhiteList(['draft', 'processing']), 'Message Status.', true)
```
Comment (abnegate) [inline app/controllers/api/messaging.php:1565]: Let's add in the description that the value must be one of either draft or status

File: app/controllers/api/messaging.php (hunk 17 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: added querying

File: app/controllers/api/messaging.php (hunk 19 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: done this

File: app/controllers/api/messaging.php (hunk 18 above)
Comment (fanatic75) [reply app/controllers/api/messaging.php]: done this

File: src/Appwrite/Utopia/Database/Validator/Queries/Subscribers.php [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+        'topicId'
+    ];
+
+    /**
+     * Expression constructor
+     *
+     */
+    public function __construct()
+    {
+        parent::__construct('messages', self::ALLOWED_ATTRIBUTES);
```
[the remark this answers is not in the document]
Comment (fanatic75) [reply src/Appwrite/Utopia/Database/Validator/Queries/Subscribers.php]: fixed this

**Review by abnegate — APPROVED** [review]

**Pull request merged**